storage/innobase/handler/handler0alter.cc

```diff
@@ -4211,103 +4211,179 @@ innobase_dropping_foreign(
+@return true if name exists or false if name doesn't exist. */
+static
+bool col_exists(unsigned n_fields, const char *name, const char **names)
+{
+  for (unsigned i = 0; i < n_fields; i++)
+  {
+    if (!strcmp(names[i], name))
+      return true;
+  }
+  return false;
+}
+
+/** Determines if an InnoDB FOREIGN KEY constraint depends on
+the nullability changes of a column.
+Enforce the following rules:
+
+i) Don't allow the child foreign key column from NULL TO NOT NULL when
+     1) Foreign key constraint type is ON UPDATE SET NULL
+     2) Foreign key constraint type is ON DELETE SET NULL
+     3) Foreign key constraint type is UPDATE CASCADE and referenced
+        column allows NULL values
+
+ii) Don't allow the parent referenced column from NOT NULL to NULL when
+foreign key constraint type is UPDATE CASCADE and foreign column
+doesn't allow NULL values
+
 @param user_table InnoDB table as it is before the ALTER operation
-@param col_name Name of the column being altered
 @param drop_fk constraints being dropped
 @param n_drop_fk number of constraints that are being dropped
-@param drop true=drop column, false=set NOT NULL
+@param col_name  modified column name
+@param new_field_flags Modified field flags
 @retval true Not allowed (will call my_error())
 @retval false Allowed
 */
-MY_ATTRIBUTE((pure, nonnull(1,4), warn_unused_result))
 static
-bool
-innobase_check_foreigns_low(
-	const dict_table_t*	user_table,
-	dict_foreign_t**	drop_fk,
-	ulint			n_drop_fk,
-	const char*		col_name,
-	bool			drop)
+bool innobase_check_foreigns_nullability(
+  const dict_table_t *user_table,
+  dict_foreign_t     **drop_fk,
+  ulint              n_drop_fk,
+  const char*        col_name,
+  uint32_t           new_field_flags)
 {
-	dict_foreign_t*	foreign;
-	ut_ad(mutex_own(&dict_sys.mutex));
-
-	/* Check if any FOREIGN KEY constraints are defined on this
-	column. */
-
-	for (dict_foreign_set::const_iterator it = user_table->foreign_set.begin();
-	     it != user_table->foreign_set.end();
-	     ++it) {
+  ut_ad(mutex_own(&dict_sys.mutex));
 
-		foreign = *it;
+  /* Changing from NOT NULL to NULL. So check referenced set */
+  if (!(new_field_flags & NOT_NULL_FLAG))
+    goto ref_field_check;
 
-		if (!drop && !(foreign->type
-			       & (DICT_FOREIGN_ON_DELETE_SET_NULL
-				  | DICT_FOREIGN_ON_UPDATE_SET_NULL))) {
-			continue;
-		}
+  for (dict_foreign_t *foreign : user_table->foreign_set)
+  {
+    if (innobase_dropping_foreign(foreign, drop_fk, n_drop_fk))
+      continue;
 
-		if (innobase_dropping_foreign(foreign, drop_fk, n_drop_fk)) {
-			continue;
-		}
+    if ((foreign->type & DICT_FOREIGN_ON_UPDATE_CASCADE) &&
+        foreign->referenced_index)
+    {
+      for (unsigned f = 0;f < foreign->n_fields; f++)
+      {
+        dict_field_t* field= dict_index_get_nth_field(
+                               foreign->referenced_index, f);
+        if (!(field->col->prtype & DATA_NOT_NULL))
+          goto col_exist_check;
+      }
+    }
```

Comment:
Between the two `}` it would be helpful to add a comment
```c++
/* All referenced columns are NOT NULL */
```
But, wouldn’t be more readable to write `foreign->referenced_index->get_n_nullable(foreign->n_fields)`? Yes, that would skip the short-circuit `goto`, but this is not time critical code.